src/Equations/Fluid/LCFSLossRateTerm.cpp

```diff
@@ -0,0 +1,208 @@
+        
+        this->psi_edge_t0 = Psihigh + slope * (Redge - Rhigh); 
+    }
+    
+    // Find if the sign of PsiDiff() should be +1 or -1
+    if(this->signFixed == false) { 
+        real_t psi_mid_init = unknowns->GetUnknownInitialData(id_psi)[0]; // Use interpolate instead?
+        real_t psi_edge_init = unknowns->GetUnknownInitialData(id_psi)[nr-1];
+        if(psi_mid_init-psi_edge_init > 0) { this->sign = 1; }
+        this->signFixed = true;
+    }
+    
+    bool exists = false;
+    for (len_t ir = 0; ir < nr; ir++){
+        if(this->sign * PsiDiff(ir) >= 0){
+            this->ir_LCFS = (int_t)ir; // Last ir to pass critera is ir_LCFS
+            exists = true;
+        }
+    }
+    if ( exists == false ) { this->ir_LCFS = -1; } // No flux surface is closed
+}
+
+
+
+/**
+* Difference between the poloidal flux at radial point ir and at the edge at t=0
+*/
+real_t LCFSLossRateTerm::PsiDiff(len_t ir){
+    //real_t psi = unknowns->GetUnknownDataPrevious(id_psi)[ir]; 
+    // Interpolate (extrapolate for ir=0) psi for estimate of psi at inner 
+    // radial grid cell wall of current grid cell
+    real_t psi = InterpolatePsi(ir); 
+    return psi - this->psi_edge_t0;
+}
+
+
+
+
+/**
+* Interpolate flux between ir:s to get psi at radial grid cell walls 
+* (Doesn't work for nr=1, add warning / if...else... with old condition in PsiDiff?)
+*/
+real_t LCFSLossRateTerm::InterpolatePsi(len_t jr){
+
+    // jr index for grid cell walls
+    len_t ir_j; // Helper index, lowest ir for which to find psi 
+    if (jr == 0) {
+        ir_j = 0; // Extrapolate for jr=0
+    } else {
+        ir_j = jr-1;
+    }
+    
+    // Need special solution if nr=1(0)?
+    real_t Rlow = this->rGrid->GetR(ir_j);
+    real_t Rhigh = this->rGrid->GetR(ir_j+1);
+    real_t Psilow = unknowns->GetUnknownDataPrevious(id_psi)[ir_j]; // Previous data to skip dependence of psi in Jacobian
+    real_t Psihigh = unknowns->GetUnknownDataPrevious(id_psi)[ir_j+1];
+    real_t slope = (Psihigh - Psilow) / (Rhigh - Rlow);
+    real_t Redge = this->rGrid->GetR_f(jr);
+    real_t interpPsi = Psilow + slope * (Redge - Rlow); // Extrapolates when jr=0 (then Redge < Rlow)
+    
+    return interpPsi;
+}
+
+
+
+
+/**
+* Currently a step function at ir_LCFS, could be changed to continuous
+*/
+real_t LCFSLossRateTerm::StepFunction(len_t ir){
+    if ((int_t)ir > this->ir_LCFS) { 
+        return 1.; 
+    } else { 
+        return 0.; 
+    }
+}
+
+
+
+/** 
+* SETWEIGHTS: Needed for DiagonalTerm, defined similar to AvalancheGrowthTerm 
+*/
+void LCFSLossRateTerm::SetWeights() {
+
+    SetGammaLoss(); // Set GammaLoss to be the weights on the radial grid
+    
+    len_t offset = 0;
+    for (len_t ir = 0; ir < nr; ir++){
+        const len_t xiIndex = this->GetXiIndexForEDirection(ir);
+        const real_t V = this->GetVolumeScaleFactor(ir);
+
+        weights[offset + n1[ir]*xiIndex] = scaleFactor * this->GammaLoss[ir] * V;
```

Comment:
Are you sure this is how you want to implement it in kinetic mode? This will just remove particles from a single point in $(p,\xi_0)$. Wouldn't it be more natural to make it uniform across all of momentum space? In that case, the term should have the same value everywhere, and that value just has to be scaled appropriately with $\mathcal{V}'\Delta p\Delta\xi_0$.